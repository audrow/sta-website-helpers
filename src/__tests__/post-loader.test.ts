import PostLoader from '../post-loader'
import {join} from 'path'
import podcast from './data/podcast.config'

describe('PostLoader init', () => {
  {
    const dataDirectory = join(__dirname, 'data', 'posts')
    it('loads all the posts with debug', async () => {
      const ph = new PostLoader(podcast, {
        isDebug: true,
        isNewestPostFirst: true,
      })
      await ph.init(dataDirectory)

      expect(ph.getPosts().length).toBe(4)
      const slugs = ph.getSlugs()
      expect(slugs.length).toBe(4)
      expect(slugs).toEqual([
        '99-future-post',
        '11-cbq',
        '10-brett-aldrich',
        '0-welcome',
      ])
    })

    it('loads all the posts in reverse order with debug', async () => {
      const ph = new PostLoader(podcast, {
        isDebug: true,
        isNewestPostFirst: false,
      })
      await ph.init(dataDirectory)

      expect(ph.getPosts().length).toBe(4)
      const slugs = ph.getSlugs()
      expect(slugs.length).toBe(4)
      expect(slugs).toEqual([
        '0-welcome',
        '10-brett-aldrich',
        '11-cbq',
        '99-future-post',
      ])
    })

    it('loads past posts without debug', async () => {
      const ph = new PostLoader(podcast, {
        isDebug: false,
        isNewestPostFirst: true,
      })
      await ph.init(dataDirectory)

      expect(ph.getPosts().length).toBe(3)
      const slugs = ph.getSlugs()
      expect(slugs.length).toBe(3)
      expect(slugs).toEqual(['11-cbq', '10-brett-aldrich', '0-welcome'])
    })

    it('creates a tag list', async () => {
      const ph = new PostLoader(podcast, {
        isDebug: false,
        isNewestPostFirst: true,
      })
      await ph.init(dataDirectory)

      expect(ph.getTags()).toMatchSnapshot()
      const slugs = ph.getSlugsByTag('startup')
      const expectedSlugs = ['11-cbq', '10-brett-aldrich']
      expect(slugs).toEqual(expect.arrayContaining(expectedSlugs))
      const posts = ph.getPosts(slugs)
      expect(posts.map((p) => p.slug)).toEqual(expectedSlugs)
    })

    it('throws when you miss init', async () => {
      const ph = new PostLoader(podcast, {
        isDebug: true,
        isNewestPostFirst: true,
      })
      expect(() => ph.getSlugs()).toThrowError()
      expect(() => ph.getPosts()).toThrowError()
      expect(() => ph.getPostBySlug('0-welcome')).toThrowError()
      await ph.init(dataDirectory)
      expect(() => ph.getSlugs()).not.toThrowError()
      expect(() => ph.getPosts()).not.toThrowError()
      expect(() => ph.getPostBySlug('0-welcome')).not.toThrowError()
    })
  }

  {
    const dataDirectory = join(__dirname, 'data', 'duplicate-slugs')
    it('throws an error when there are duplicate slugs', async () => {
      const ph = new PostLoader(podcast, {
        isDebug: true,
        isNewestPostFirst: true,
      })
      expect(ph.init(dataDirectory)).rejects.toThrowError()
    })
  }
})

describe('PostLoader getPostBySlug', () => {
  const dataDirectory = join(__dirname, 'data', 'posts')
  it('gets a post with a number by slug', async () => {
    const ph = new PostLoader(podcast, {
      isDebug: false,
      isNewestPostFirst: true,
    })
    await ph.init(dataDirectory)
    const post = ph.getPostBySlug('11-cbq')
    expect(post.slug).toBe('11-cbq')
    expect(post).toMatchSnapshot()
  })
  it('gets a post with a custom slug by slug', async () => {
    const ph = new PostLoader(podcast, {
      isDebug: false,
      isNewestPostFirst: true,
    })
    await ph.init(dataDirectory)
    const post = ph.getPostBySlug('0-welcome')
    expect(post.slug).toBe('0-welcome')
    expect(post).toMatchSnapshot()
  })
  it('throws an error when there is no matching slug', async () => {
    const ph = new PostLoader(podcast, {
      isDebug: false,
      isNewestPostFirst: true,
    })
    await ph.init(dataDirectory)
    expect(() => ph.getPostBySlug('not-a-slug')).toThrowError()
  })
})
