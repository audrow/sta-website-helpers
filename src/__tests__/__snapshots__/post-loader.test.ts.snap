// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostLoader getPostBySlug gets a post with a custom slug by slug 1`] = `
Object {
  "customMp3FileName": "STA Ep 0 - Welcome",
  "customSlug": "0-welcome",
  "description": "Audrow Nash introduces the Sense Think Act Podcast.",
  "duration": Object {
    "hours": 0,
    "minutes": 4,
    "seconds": 39,
  },
  "excerpt": "Audrow Nash introduces the Sense Think Act Podcast.",
  "mp3": Object {
    "sizeBytes": 9097894,
    "url": "https://ftp.osuosl.org/pub/ros/download.ros.org/sensethinkact/episodes/STA%20Ep%200%20-%20Welcome.mp3",
  },
  "mp3SizeBytes": 9097894,
  "slug": "0-welcome",
  "tags": Array [
    "announcement",
  ],
  "title": "Welcome to Sense Think Act",
  "url": "https://sensethinkact.com/episodes/0-welcome",
  "youtube": Object {
    "mainContentId": "welcome",
  },
}
`;

exports[`PostLoader getPostBySlug gets a post with a number by slug 1`] = `
Object {
  "description": "In this episode, Audrow Nash speaks to Charles Brian Quinn (aka, CBQ), CEO and a Co-Founder of Greenzie. Greenzie make an autonomous driving system for commercial lawn mowers. We talk about Greenzie's autonomous mowing system, how Greenzie has worked with manufacturers to up-fit their system into comercial mowers, how Greenzie does dog-fooding, safety and standards, and about CBQ's experiences bootstrapping and with venture capital.",
  "duration": Object {
    "hours": 1,
    "minutes": 51,
    "seconds": 46,
  },
  "excerpt": "CBQ is the CEO and a Co-Founder of Greenzie.",
  "guests": "CBQ",
  "links": Array [
    Object {
      "name": "CBQ's website",
      "url": "https://www.seebq.com/",
    },
    Object {
      "name": "Greenzie's website",
      "url": "https://www.greenzie.com/",
    },
    Object {
      "name": "Greenzie's YouTube channel",
      "url": "https://www.youtube.com/c/Greenzie",
    },
  ],
  "mp3": Object {
    "sizeBytes": 225870038,
    "url": "https://ftp.osuosl.org/pub/ros/download.ros.org/sensethinkact/episodes/STA%20Ep%2011%20-%20CBQ.mp3",
  },
  "mp3SizeBytes": 225870038,
  "number": 11,
  "slug": "11-cbq",
  "tags": Array [
    "startup",
    "software",
    "sensor fusion",
    "autonomous driving",
    "mowers",
    "safety",
    "dogfooding",
  ],
  "title": "Commercial-grade Autonomous Mowers, Safety, and Dogfooding",
  "url": "https://sensethinkact.com/episodes/11-cbq",
  "youtube": Object {
    "clips": Array [
      Object {
        "baseTitle": "Does it Detect Sprinkler Heads?",
        "videoId": "ahUd_2Uydps",
      },
      Object {
        "baseTitle": "Making Perfect Turns",
        "videoId": "odAvY-ojVAA",
      },
      Object {
        "baseTitle": "Seasonal Workers",
        "videoId": "6CfAz0z07I4",
      },
      Object {
        "baseTitle": "Leap of Faith",
        "videoId": "8oVBa43dQGw",
      },
      Object {
        "baseTitle": "Differences Between Kids and Poles",
        "videoId": "HrraVxtyxok",
      },
    ],
    "mainContentId": "pBWLSmoBj54",
  },
}
`;

exports[`PostLoader init creates a tag list 1`] = `
Array [
  "announcement",
  "autonomous driving",
  "behavior trees",
  "control",
  "dogfooding",
  "mowers",
  "safety",
  "sensor fusion",
  "software",
  "startup",
  "state machines",
]
`;
